drv/fpga-devices/src/ecp5_spi_mux_pca9538.rs

```diff
@@ -0,0 +1,297 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::ecp5::{Command, Ecp5Driver};
+use crate::FpgaUserDesign;
+use drv_fpga_api::FpgaError;
+use drv_i2c_api::ResponseCode;
+use drv_i2c_devices::pca9538;
+use drv_spi_api::{self as spi_api, SpiDevice, SpiError};
+use drv_stm32xx_sys_api::{self as sys_api, GpioError, Sys};
+
+/// This module implements an ECP5 driver which exposes two physical devices,
+/// which share a single SPI bus using a mux and are controlled through a shared
+/// PCA9538 GPIO expander.
+
+/// Impl Error type, with conversion from `GpioError`, `SpiError` and
+/// `ResponseCode`.
+#[derive(Copy, Clone, Debug)]
+pub enum Error {
+    GpioError(GpioError),
+    SpiError(SpiError),
+    I2cError(ResponseCode),
+}
+
+impl From<GpioError> for Error {
+    fn from(e: GpioError) -> Self {
+        Self::GpioError(e)
+    }
+}
+
+impl From<SpiError> for Error {
+    fn from(e: SpiError) -> Self {
+        Self::SpiError(e)
+    }
+}
+
+impl From<ResponseCode> for Error {
+    fn from(e: ResponseCode) -> Self {
+        Self::I2cError(e)
+    }
+}
+
+impl From<Error> for u8 {
+    fn from(e: Error) -> Self {
+        match e {
+            Error::GpioError(e) => match e {
+                GpioError::BadArg => 2,
+            },
+            Error::SpiError(e) => match e {
+                SpiError::BadTransferSize => 3,
+                SpiError::ServerRestarted => 4,
+                SpiError::NothingToRelease => 5,
+                SpiError::BadDevice => 6,
+                SpiError::DataOverrun => 7,
+            },
+            Error::I2cError(e) => 8 + (e as u8),
+        }
+    }
+}
+
+impl From<Error> for FpgaError {
+    fn from(e: Error) -> Self {
+        FpgaError::ImplError(u8::from(e))
+    }
+}
+
+pub struct Pins {
+    pub done: pca9538::PinSet,
+    pub init_n: pca9538::PinSet,
+    pub program_n: pca9538::PinSet,
+    pub user_design_reset_n: pca9538::PinSet,
+}
+
+pub struct Driver {
```

Comment:
I don't love having every member `pub` here – since there's a cache in `instance_select`, it makes it possible to invalidate state.  However, I _do_ like having the constructor specify variables by name.

How about making internal variables of the `Driver` private, then adding
```
struct DriverConfig {
    pub sys: Sys,
    pub configuration_port: SpiDevice,
    pub user_design: SpiDevice,
    pub spi_mux_select: sys_api::PinSet,
    pub gpio: pca9538::Pca9538,
    pub user_design_reset_duration: u64,
}
```
with an initialization function
```
impl Driver {
    pub fn init(config: DriverConfig) -> Self {
       ...
    }
}
```
The `init` function should also does GPIO configuration to ensure that the `instance_selected` is correct (currently done in `configure_gpio`).  That way, the only way to get a `struct Driver` ensures that it's valid.

Reply:
Yup, this split is a good idea. It's a common pattern I use in Bluespec code, not sure why I didn't think of it here.

Reply:
I've refactored this quite a bit and overall I am pretty happy with the result. The only thing that's nagging a bit is the `config` member in `Driver` causing deeper nesting elsewhere in the driver. But the immediate alternative is copying the fields from `DriverConfig` to `Driver`, which seemed equally meh. I was contemplating something with `Deref`, but that didn't look that great either. Any suggestions on something I have not thought of?

Reply:
Yeah, it seems like all the options are meh:
- Storing the `DriverConfig` in the driver adds an extra `.config` all over
- Copying fields is a lot of duplication
- Using `Deref` for anything other than a smart pointer is heavily frowned upon

I think options 1 vs 2 are about the same, so no objections to the one you picked.